### PR TITLE
Add ParamSpec bindings

### PR DESCRIPTION
**Pull request by sdroege**

### CODE REVIEW

File: src/gobject/auto/param_spec.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+use object::IsA;
+use std::mem;
+use std::ptr;
+use translate::*;
+
+glib_wrapper! {
+    pub struct ParamSpec(Object<ffi::GParamSpec, ffi::GParamSpecClass>);
+
+    match fn {
+        get_type => || ffi::intern(),
```
Comment (sdroege) [inline src/gobject/auto/param_spec.rs]: We need a way to override the get_type functions. This one here has a constant, `G_TYPE_PARAMSPEC`

**Comment by sdroege** [comment]
For some reason things like `g_param_spec_bool`, `g_param_spec_int`, etc (all the constructors) are not generated by GIR. Not sure yet why, @EPashkin do you have an idea?

Also how would we best override the `get_type` function? Configuration that takes a string, and that is just put there instead of whatever the GIR says?

File: Gir_GObject.toml [hunk 2]
```diff
@@ -11,15 +11,42 @@ single_version_file = true
...
 [[object]]
 name = "GObject.*"
 status = "ignore"
+
+[[object]]
+name = "GObject.ParamSpec"
+status = "generate"
+    [[object.function]]
+    name = "sink"
+    ignore = true # memory management
```
Comment (EPashkin) [inline Gir_GObject.toml]: Please move comment to previous line

File: src/gobject/auto/param_spec.rs (hunk 1 above)
Comment (EPashkin) [reply src/gobject/auto/param_spec.rs]: Agree

**Comment by EPashkin** [comment]
IMHO config string  `get_type: Option<String>` is enough

**Comment by sdroege** [comment]
While at it, I would add the same for ref/copy/unref/free functions the same way. That allows me to get rid of some more manual bindings

**Comment by EPashkin** [comment]
Thanks in advance

File: src/gobject/auto/param_spec.rs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+use gobject_ffi as ffi;
+use ffi as glib_ffi;
+use gobject_ffi;
+use object::IsA;
+use std::mem;
+use std::ptr;
+use translate::*;
+
+glib_wrapper! {
+    pub struct ParamSpec(Object<ffi::GParamSpec, ffi::GParamSpecClass>);
```
Comment (sdroege) [inline src/gobject/auto/param_spec.rs]: This is actually not a `GObject` but another fundamental type. So I guess we don't really get around manually binding this one after all...

Will update the PR accordingly

**Comment by sdroege** [comment]
Updated accordingly and made use of it in `object.rs` a bit

File: src/object.rs [hunk 4]
```diff
@@ -673,33 +675,32 @@ impl<T: IsA<Object> + SetValue> ObjectExt for T {
...
-            let object = self.to_glib_none().0;
-            let klass = (*object).g_type_instance.g_class as *mut gobject_ffi::GObjectClass;
-            let pspec = gobject_ffi::g_object_class_find_property(klass, property_name.to_glib_none().0);
-            if pspec.is_null() {
+        let pspec = match self.find_property(property_name) {
+            Some(pspec) => pspec,
+            None => {
                 return Err(BoolError("property not found"));
-            }
+            },
```
Comment (EPashkin) [inline src/object.rs]: comma unneeded

File: src/object.rs [hunk 5]
```diff
@@ -673,33 +675,32 @@ impl<T: IsA<Object> + SetValue> ObjectExt for T {
...
+        };
 
-            let writeable = (*pspec).flags & gobject_ffi::G_PARAM_WRITABLE != 0;
-            if !writeable {
-                return Err(BoolError("property is not writeable"));
-            }
+        if !pspec.get_flags().contains(::ParamFlags::WRITABLE) {
```
Comment (EPashkin) [inline src/object.rs]: Maybe flags need check for `CONSTRUCT_xxx`, some properties have WRITABLE and CONSTRUCT_ONLY, so actually not writable

Comment (sdroege) [reply src/object.rs]: Thanks, good idea :) Will change in a few minutes

File: src/object.rs (hunk 4 above)
Comment (sdroege) [reply src/object.rs]: done

File: src/object.rs (hunk 5 above)
Comment (sdroege) [reply src/object.rs]: done

File: src/object.rs (hunk 4 above)
Comment (EPashkin) [reply src/object.rs]: Strange, I still see it, and in other `match self.find_property` down too.
Seems you forgot to push 😉

Comment (sdroege) [reply src/object.rs]: I indeed forgot to push it, yes :) Thanks

**Comment by sdroege** [comment]
All green :)

**Comment by EPashkin** [comment]
👍 then

File: src/param_spec.rs [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        }
+    }
+
+    pub fn value_array(name: &str, nick: &str, blurb: &str, element_spec: &ParamSpec, flags: ParamFlags) -> Option<ParamSpec> {
+        unsafe {
+            from_glib_none(ffi::g_param_spec_value_array(name.to_glib_none().0, nick.to_glib_none().0, blurb.to_glib_none().0, element_spec.to_glib_none().0, flags.to_glib()))
+        }
+    }
+
+
```
Comment (GuillaumeGomez) [inline src/param_spec.rs]: Extra empty line.

File: src/param_spec.rs [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    pub fn variant<'a, P: Into<Option<&'a ::Variant>>>(name: &str, nick: &str, blurb: &str, type_: &::VariantTy, default_value: P, flags: ParamFlags) -> Option<ParamSpec> {
+        unsafe {
+            let default_value = default_value.into();
+            from_glib_none(ffi::g_param_spec_variant(name.to_glib_none().0, nick.to_glib_none().0, blurb.to_glib_none().0, type_.to_glib_none().0, default_value.to_glib_none().0, flags.to_glib()))
+        }
+    }
+
```
Comment (GuillaumeGomez) [inline src/param_spec.rs]: Extra empty line.

File: src/param_spec.rs (hunk 6 above)
Comment (sdroege) [reply src/param_spec.rs]: fixed

File: src/param_spec.rs (hunk 7 above)
Comment (sdroege) [reply src/param_spec.rs]: fixed

**Comment by sdroege** [comment]
And green again :)

**Comment by GuillaumeGomez** [comment]
Thanks!

**Pull request merged**